RecoLocalCalo/HcalRecAlgos/src/ZdcSimpleRecAlgo_Run3.cc

```diff
@@ -0,0 +1,547 @@
+      // If time is zero, set it to the "nonsensical" -99
+      // Time should be between 75ns and 175ns (Timeslices 3-7)
+      if (AvgTSPos == 0) {
+        time = -99;
+      } else {
+        time = (AvgTSPos * 25.0);
+      }
+    }
+
+    // find energy for signal TS + 1
+    for (unsigned int ivs = 0; ivs < mySignalTS.size(); ++ivs) {
+      CurrentTS = mySignalTS[ivs] + 1;
+      if (CurrentTS >= digi_size)
+        continue;
+      int capid = digi[CurrentTS].capid();
+      // ta = tool[CurrentTS] - noise;
+      ta = tool[CurrentTS];
+      ta *= calibs.respcorrgain(capid);  // fC --> GeV
+      if (ta > 0)
+        energySOIp1 += ta;
+    }
+
+    double tmp_energy = 0;
+    double tmp_TSWeightedEnergy = 0;
+    for (int ts = 0; ts < 6; ++ts) {
+      if (CurrentTS >= digi_size)
+        continue;
+      int capid = digi[ts].capid();
+
+      // max sure there are no negative values in time calculation
+      ta = std::max(0.0, tool[ts]);
+      ta *= calibs.respcorrgain(capid);  // fC --> GeV
+      if (ta > 0) {
+        tmp_energy += ta;
+        tmp_TSWeightedEnergy += (ts + 1) * ta;
+      }
+    }
+
+    chargeWeightedTime = (tmp_TSWeightedEnergy / tmp_energy - 1) * 25.0;
+    auto rh = RecHit(digi.id(), ampl, time, lowGEnergy);
+    rh.setEnergySOIp1(energySOIp1);
+
+    if (maxI >= 0 && maxI < tool.size()) {
+      float tmp_tdctime = 0;
+      // TDC error codes will be 60=-1, 61 = -2, 62 = -3, 63 = -4
+      if (digi[maxI].le_tdc() >= 60)
+        tmp_tdctime = -1 * (digi[maxI].le_tdc() - 59);
+      else
+        tmp_tdctime = maxI * 25. + (digi[maxI].le_tdc() / 2);
+      rh.setTDCtime(tmp_tdctime);
+    }
+
+    rh.setChargeWeightedTime(chargeWeightedTime);
+    rh.setRatioSOIp1(ratioSOIp1);
+    return rh;
+  }
+}  // namespace ZdcSimpleRecAlgoImpl
+
+//reco2 method should make the Energy Assosciated with the ZDC Trigger Energy as of June 2024
+// only difference between this method and the trigger value LUT value is the signal and noise TS
+// are bins into 1024 bins of 50 GeV bin widths. i.e. energyint = std::min(std::max(0, int(energy/(50))), 1023)
+// this assumes the noise TS =2 and signalTS =2
+namespace ZdcSimpleRecAlgoImpl {
+  template <class Digi, class RecHit>
+  inline RecHit reco2(const Digi& digi,
+                      const HcalCoder& coder,
+                      const HcalCalibrations& calibs,
+                      const HcalPedestal& effPeds,
+                      const std::vector<unsigned int>& myNoiseTS,
+                      const std::vector<unsigned int>& mySignalTS) {
+    CaloSamples tool;
+    coder.adc2fC(digi, tool);
+    // Reads noiseTS and signalTS from database
+    int ifirst = mySignalTS[0];
+    //    int n = mySignalTS.size();
+    double ampl = 0;
+    int maxI = -1;
+    double maxA = -1e10;
+    double ta = 0;
+    double lowGEnergy = 0;
+    double energySOIp1 = 0;
+    double ratioSOIp1 = -1;
+    double chargeWeightedTime = 0;
+
+    double noiseFrac = 97.0 / 256.0;
```

Comment:
Please explain with a comment.

Reply:
code has been updated and now the noiseFrac is now part of the cfi file and is now called ootpuFrac. The ootpuFrac is fraction of energy from the previous timeslice that is expected to contribute to the charge of the signal timeslice. This is expected to be about .4 in events with pre pileup.